internal/poolutil/pool_test.go

```diff
@@ -0,0 +1,152 @@
+package poolutil
+
+import (
+	"inet.af/netaddr"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("FindFreeAddress", func() {
+	var (
+		freeIPSet *netaddr.IPSet
+		existing  *netaddr.IPSet
+	)
+
+	type iprange struct {
+		From string
+		To   string
+	}
+
+	createIPSet := func(ipranges []iprange) *netaddr.IPSet {
+		builder := netaddr.IPSetBuilder{}
+		for _, r := range ipranges {
+			first, err := netaddr.ParseIP(r.From)
+			Expect(err).NotTo(HaveOccurred())
+
+			second, err := netaddr.ParseIP(r.To)
+			Expect(err).NotTo(HaveOccurred())
+
+			builder.AddRange(netaddr.IPRangeFrom(first, second))
+		}
+
+		ipSet, err := builder.IPSet()
+		Expect(err).NotTo(HaveOccurred())
+
+		return ipSet
+	}
+
+	BeforeEach(func() {
+		freeIPSet = nil
+
+		var err error
+		existingIPSetBuilder := netaddr.IPSetBuilder{}
+		existing, err = existingIPSetBuilder.IPSet()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	When("there is a single IPRange", func() {
+		BeforeEach(func() {
+			freeIPSet = createIPSet([]iprange{
+				{From: "10.0.0.10", To: "10.0.0.20"},
+			})
+		})
+
+		It("returns the first address in that range", func() {
+			ip, err := FindFreeAddress(freeIPSet, existing)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(ip.String()).To(Equal("10.0.0.10"))
+		})
+
+		When("there are some addresses claimed", func() {
+			BeforeEach(func() {
+				existing = createIPSet([]iprange{
+					{From: "10.0.0.10", To: "10.0.0.11"},
+				})
+			})
+
+			It("returns a free address from that range", func() {
+				ip, err := FindFreeAddress(freeIPSet, existing)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(ip.String()).To(Equal("10.0.0.12"))
+			})
+		})
+
```

Comment:
Since I wasn't sure that this is working while looking at the diff - could you add this test here?

```go
    When("only the last ip address is available", func() {
			BeforeEach(func() {
				existing = createIPSet([]iprange{
					{From: "10.0.0.10", To: "10.0.0.19"},
				})
			})

			It("returns that ip address", func() {
				ip, err := FindFreeAddress(freeIPSet, existing)
				Expect(err).NotTo(HaveOccurred())
				Expect(ip.String()).To(Equal("10.0.0.20"))
			})
		})
```

Reply:
Good eye!

Reply:
I've trained it 😄 https://github.com/telekom/cluster-api-ipam-provider-in-cluster/commit/e0659c80f118a3ee360bb945eee2b7ccf019b4ec